src/ovm/ObolValidatorManager.sol

```diff
@@ -0,0 +1,509 @@
+    // Check if fee exceeds maximum allowed, otherwise get fee
+    uint256 fee = _validateAndReturnFee(consolidationSystemContract, maxFeePerConsolidation);
+
+    // Calculate total number of consolidation operations
+    uint256 totalNumOfConsolidationOperations = 0;
+    for (uint256 i = 0; i < requests.length; i++) {
+      if (requests[i].srcPubKeys.length == 0 || requests[i].srcPubKeys.length > 63) {
+        revert InvalidRequest_Params();
+      }
+      totalNumOfConsolidationOperations += requests[i].srcPubKeys.length;
+    }
+    // Check if the msg.value is enough to cover the fees
+    uint256 totalFeeRequired = fee * totalNumOfConsolidationOperations;
+    _validateSufficientValueForFee(msg.value, totalFeeRequired);
+
+    // Perform the consolidation requests
+    for (uint256 i = 0; i < requests.length; i++) {
+      _validatePubkeyLength(requests[i].targetPubKey);
+
+      for (uint256 j = 0; j < requests[i].srcPubKeys.length; j++) {
+        _validatePubkeyLength(requests[i].srcPubKeys[j]);
+
+        // Add the consolidation request
+        bytes memory callData = bytes.concat(requests[i].srcPubKeys[j], requests[i].targetPubKey);
+        (bool success, ) = consolidationSystemContract.call{value: fee}(callData);
+        if (!success) {
+          revert InvalidConsolidation_Failed();
+        }
+        
+        // Emit consolidation event for each operation
+        emit ConsolidationRequested(requests[i].srcPubKeys[j], requests[i].targetPubKey, fee);
+      }
+    }
+    // Refund any excess value back to the excessFeeRecipient
+    _refundExcessFee(msg.value, totalFeeRequired, excessFeeRecipient);
+  }
+
+  /// @inheritdoc IObolValidatorManager
+  function withdraw(
+    bytes[] calldata pubKeys,
+    uint64[] calldata amounts,
+    uint256 maxFeePerWithdrawal,
+    address excessFeeRecipient
+  ) external payable onlyOwnerOrRoles(WITHDRAWAL_ROLE) nonReentrant {
+    if (pubKeys.length != amounts.length) revert InvalidRequest_Params();
+
+    // check if the value sent is enough to cover the fees
+    _validateSufficientValueForFee(msg.value, maxFeePerWithdrawal * pubKeys.length);
+
+    // Check if fee exceeds maximum allowed, otherwise get fee
+    uint256 fee = _validateAndReturnFee(withdrawalSystemContract, maxFeePerWithdrawal);
+    uint256 totalFeePaid = 0;
+    
+    for (uint256 i; i < pubKeys.length; i++) {
+      _validatePubkeyLength(pubKeys[i]);
+
+      // Add the withdrawal request
+      bytes memory callData = abi.encodePacked(pubKeys[i], amounts[i]);
+      (bool success, ) = withdrawalSystemContract.call{value: fee}(callData);
+      if (!success) {
+        revert InvalidWithdrawal_Failed();
+      }
+      totalFeePaid += fee;
+
+      // Emit withdrawal event for each validator
+      emit WithdrawalRequested(pubKeys[i], amounts[i], fee);
+    }
+
+    // Refund any excess value back to the excessFeeRecipient
+    _refundExcessFee(msg.value, totalFeePaid, excessFeeRecipient);
```

Comment:
future improvement idea is to figure out if msg.value - totalFeePaid > eth_call * gas price. Might often be a wasteful endeavour, idk.